File: 89.R
# Roman numerals

# Constants
ROMAN_URL <- 'https://projecteuler.net/project/resources/p089_roman.txt'
TO_DECIMAL <- c(1, 5, 10, 50, 100, 500, 1000)
names(TO_DECIMAL) <- c('I', 'V', 'X', 'L', 'C', 'D', 'M')
TO_ROMAN <- c(1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1)
names(TO_ROMAN) <- c('M', 'CM', 'D', 'CD', 'C', 'XC', 'L', 'XL', 'X', 'IX', 'V',
                     'IV', 'I')

to_decimal <- function(roman) {
    decimal <- 0

    while ( nchar(roman) ) {
        if ( grepl('^(IV|IX|XL|XC|CD|CM)', roman) ) {
            num1 <- substring(roman, 1, 1)
            num2 <- substring(roman, 2, 2)
            roman <- substring(roman, 3)
            decimal <- decimal + TO_DECIMAL[num2] - TO_DECIMAL[num1]
        } else {
            num <- substring(roman, 1, 1)
            roman <- substring(roman, 2)
            decimal <- decimal + TO_DECIMAL[num]
        }
    }

    return(decimal)
}

to_roman <- function(decimal) {
    roman <- vector()

    for ( i in seq.int(length(TO_ROMAN)) ) {
        quotient <- decimal %/% TO_ROMAN[i]
        if ( quotient ) {
            for ( j in seq.int(quotient) ) {
                roman <- c(roman, names(TO_ROMAN)[i])
            }
        }
        decimal <- decimal %% TO_ROMAN[i]
    }
    roman <- paste(roman, collapse='')

    return(roman)
}

non_minimal_numbers <- suppressWarnings(read.delim(sub('https', 'http',
                                                       ROMAN_URL), header=FALSE,
                                        colClasses='character'))[,1]

minimal_numbers <- vector()
for ( number in non_minimal_numbers ) {
    minimal_numbers <- c(minimal_numbers, to_roman(to_decimal(number)))
}
non_minimal_length <- sum(nchar(as.character(non_minimal_numbers)))
minimal_length <- sum(nchar(as.character(minimal_numbers)))

cat(non_minimal_length - minimal_length, fill=TRUE)

File: 89.js
#!/usr/bin/env node

// Roman numerals

var ROMAN_URL = 'https://projecteuler.net/project/resources/p089_roman.txt';
var TO_DECIMAL = {
    'I': 1,
    'V': 5,
    'X': 10,
    'L': 50,
    'C': 100,
    'D': 500,
    'M': 1000
};
var TO_ROMAN = [
    [1000, 'M'],
    [900, 'CM'],
    [500, 'D'],
    [400, 'CD'],
    [100, 'C'],
    [90, 'XC'],
    [50, 'L'],
    [40, 'XL'],
    [10, 'X'],
    [9, 'IX'],
    [5, 'V'],
    [4, 'IV'],
    [1, 'I']
];

var request = require('request');
request(ROMAN_URL, function(error, response, body) {
    if (!error && response.statusCode == 200) {
        reduce_numerals(body.trim().split(/[\r\n]+/));
    }
});

function reduce_numerals(non_minimal_nums) {
    var minimal_length = 0;
    var non_minimal_length = 0;
    for (var i = 0; i < non_minimal_nums.length; i++) {
        non_minimal_length += non_minimal_nums[i].toString().length;
        var minimal_num = to_roman(to_decimal(non_minimal_nums[i]));
        minimal_length += minimal_num.toString().length;
    }

    console.log(non_minimal_length - minimal_length);
}

function to_decimal(roman) {
    var decimal = 0;

    while (roman.length) {
        if (/^(IV|IX|XL|XC|CD|CM)/.test(roman)) {
            var num1 = roman.substring(0, 1);
            var num2 = roman.substring(1, 2);
            roman = roman.substring(2);
            decimal += TO_DECIMAL[num2] - TO_DECIMAL[num1];
        } else {
            var num = roman.substring(0, 1);
            roman = roman.substring(1);
            decimal += TO_DECIMAL[num];
        }
    }

    return decimal;
}

function to_roman(decimal) {
    var roman = '';

    for (var i = 0; i < TO_ROMAN.length; i++) {
        var quotient = Math.floor(decimal / TO_ROMAN[i][0]);
        for (var j = 0; j < quotient; j++) {
            roman += TO_ROMAN[i][1];
        }
        decimal = decimal % TO_ROMAN[i][0];
    }

    return roman;
}
